### PR TITLE
sys/ztimer: expose ZTIMER_MSEC/USEC_BASE 

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

`ztimer` comes with a lot of flexibility, specially with the option to chain `ztimer_clock_t`. If I would want I can configure timers with different resolutions.

But since `ZTIMER_MSEC` or `ZTIMER_USEC` can themselves have been already shifted/frac/divided to get the desired frequency it makes sense to branch new timers from the base timer instead of chaining them, that way conversions are not as well chained.

To allow that this PR exposes the `BASE` timer as well as its frequency. The base frequency assumes that the timer is running at `1Mhz` or at `XTIMER_HZ` and that `RTT` is used when possible for `ZTIMER_MSEC`.

I wanted to expose the base for `ZTIMER_MSEC` or `ZTIMER_USEC` since the api expects these to be available.

It might make sense instead of exposing a `ZTIMER_BASE_LPM` (low power mode capabilities) and `ZTIMER_BASE`, since if there is no timer that can sleep, the base for `ZTIMER_MSEC` is likely to be the same than for `ZTIMER_USEC`.

### Testing procedure

- `tests/ztimer_%`still pass

### CODE REVIEW

**Comment by fjmolinas** [comment]
@jue89 might have some thoughts as well since he has been working with `ztimer`

File: sys/ztimer/Makefile.dep [hunk 1]
```diff
@@ -74,4 +74,8 @@ endif
 
 ifneq (,$(filter ztimer_msec,$(USEMODULE)))
   USEMODULE += ztimer
+  FEATURES_OPTIONAL += periph_rtt
+  ifneq (,$(filter periph_rtt,$(FEATURES_USED)))
+    USEMODULE += ztimer_periph_rtt
+  endif
```
Comment (jue89) [inline sys/ztimer/Makefile.dep:80]: I like this approach! It's heading more in the direction of *it simply works*.

File: sys/ztimer/auto_init.c [hunk 2]
```diff
@@ -107,18 +108,20 @@ ztimer_clock_t *const ZTIMER_USEC = &_ztimer_convert_frac_usec.super.super;
 #if MODULE_ZTIMER_MSEC
 #  if MODULE_PERIPH_RTT
```
Comment (jue89) [inline sys/ztimer/auto_init.c:109]: I would change this to `#if MODULE_ZTIMER_PERIPH_RTT`. This should be the correct way to check whether `ZTIMER_MSEC` should be backed by `periph_rtt`.

File: sys/include/ztimer.h [hunk 3]
```diff
@@ -488,11 +491,43 @@ void ztimer_update_head_offset(ztimer_clock_t *clock);
...
 extern ztimer_clock_t *const ZTIMER_USEC;
 
+/**
+ * @def     ZTIMER_USEC_BASE_FREQ
+ *
+ * @brief   The frequency of the base ztimer for ZTIMER_USEC
+ */
+#ifdef XTIMER_HZ
+#define ZTIMER_USEC_BASE_FREQ   (XTIMER_HZ)
+#else
```
Comment (jue89) [inline sys/include/ztimer.h:511]: On the first glimpse, this looks like `xtimer` configuration is creeping into `ztimer`. But I don't have any idea how to implement this the *clean way*.

File: sys/ztimer/Makefile.dep (hunk 1 above)
Comment (fjmolinas) [reply sys/ztimer/Makefile.dep:80]: I think I'll split the first two commits out of this one!

File: sys/include/ztimer.h [hunk 4]
```diff
@@ -488,11 +491,43 @@ void ztimer_update_head_offset(ztimer_clock_t *clock);
...
+ * @brief   Base ztimer for the millisecond clock
+ */
+extern ztimer_clock_t *const ZTIMER_MSEC_BASE;
+
+/**
+ * @def     ZTIMER_MSEC_BASE_FREQ
+ *
+ * @brief   The frequency of the base ztimer for ZTIMER_MSEC
+ */
+#ifdef MODULE_PERIPH_RTT
```
Comment (jue89) [inline sys/include/ztimer.h:525]: This should also be `#ifdef MODULE_ZTIMER_PERIPH_RTT` to be congruent with the change in `ztimer/auto_init.c` line 108.

File: sys/ztimer/auto_init.c (hunk 2 above)
Comment (jue89) [reply sys/ztimer/auto_init.c]: After a closer look, I would say this change is more related to #13967.

File: sys/include/ztimer.h [hunk 5]
```diff
@@ -488,11 +491,43 @@ void ztimer_update_head_offset(ztimer_clock_t *clock);
...
  * @brief   Default ztimer microsecond clock
  */
 extern ztimer_clock_t *const ZTIMER_USEC;
 
+/**
+ * @def     ZTIMER_USEC_BASE_FREQ
+ *
+ * @brief   The frequency of the base ztimer for ZTIMER_USEC
+ */
+#ifdef XTIMER_HZ
```
Comment (kaspar030) [inline sys/include/ztimer.h:509]: I see an issue here. I tried to keep all of the ztimer configuration in auto_init.c.
That means that the actual ztimer code compiles from the same headers. Including ztimer.h leads to the same on every platform.
Pulling the configuration here messes with that principle.

A solution would be to change the frequency defines to extern uint32 declarations, and define them in auto_init.c.

Comment (fjmolinas) [reply sys/include/ztimer.h:509]: I didn't use an `extern uint32_t` because then you can't do compile checks in other places, e.g. I can't know if I need to `shift` or scale down the timer at compile time, I would have to have a `ztimer_convert_shift_t ` and `ztimer_convert_frac_t` and at runtime check the speed to know which one to use.

Comment (fjmolinas) [reply sys/include/ztimer.h:509]: > I see an issue here. I tried to keep all of the ztimer configuration in auto_init.c.
> That means that the actual ztimer code compiles from the same headers. Including ztimer.h leads to the same on every platform.

Maybe we can have a configuration header? Similar to what we do for drivers with `driveXXX_params.h`?

Comment (kaspar030) [reply sys/include/ztimer.h:509]: fine with me, as long as from ztimer itself, only auto_init.c includes it.

Comment (kaspar030) [reply sys/include/ztimer.h:509]: `include/ztimer/config.h`?

File: sys/include/ztimer.h [hunk 6]
```diff
@@ -503,11 +503,21 @@ static inline void ztimer_init_extend(ztimer_clock_t *clock)
 }
 
 /* default ztimer virtual devices */
+/**
```
Comment (kaspar030) [inline sys/include/ztimer.h:506]: just a suggestion: when reading this, would it make sense to have ZTIMER_USEC and ZTIMER_MSEC both on top and the BASE below?

Comment (fjmolinas) [reply sys/include/ztimer.h:506]: Do you mean:

A)
```
/**
 * @brief   Default ztimer microsecond clock
 */
extern ztimer_clock_t *const ZTIMER_USEC;

/* default ztimer virtual devices */
/**
 * @brief   Base ztimer for the microsecond clock
 */
extern ztimer_clock_t *const ZTIMER_USEC_BASE;

/**
 * @brief   Default ztimer millisecond clock
 */
extern ztimer_clock_t *const ZTIMER_MSEC;

/**
 * @brief   Base ztimer for the millisecond clock
 */
extern ztimer_clock_t *const ZTIMER_MSEC_BASE;
```

or B):

```
/**
 * @brief   Default ztimer microsecond clock
 */
extern ztimer_clock_t *const ZTIMER_USEC;

/**
 * @brief   Default ztimer millisecond clock
 */
extern ztimer_clock_t *const ZTIMER_MSEC;

/* default ztimer virtual devices */
/**
 * @brief   Base ztimer for the microsecond clock
 */
extern ztimer_clock_t *const ZTIMER_USEC_BASE;

/**
 * @brief   Base ztimer for the millisecond clock
 */
extern ztimer_clock_t *const ZTIMER_MSEC_BASE;
```

File: sys/include/ztimer/config.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+#define CONFIG_ZTIMER_USEC_MIN          (10)
+#endif
+
+/**
+ * @brief   ZTIMER_USEC base timer width
+ */
+#ifndef CONFIG_ZTIMER_USEC_WIDTH
+#  if (TIMER_0_MAX_VALUE) == 0xFFFF
+#    define CONFIG_ZTIMER_USEC_WIDTH        (16)
+#    define XTIMER_WIDTH (24)
```
Comment (kaspar030) [inline sys/include/ztimer/config.h:87]: why is this 24? isn't xtimer_width always 32 when ztimer is used?

Comment (fjmolinas) [reply sys/include/ztimer/config.h:87]: copy pasta

Comment (fjmolinas) [inline sys/include/ztimer/config.h:87]:
```suggestion

```

File: sys/include/ztimer.h (hunk 6 above)
Comment (kaspar030) [reply sys/include/ztimer.h:506]: B! and maybe beef up docs that those are the clocks that the others are based on, if converted, or the same if not converted, and that the frequency of them are available as ZTIMER_X_FREQ in ztimer/config.h. I offer beers as bribe!

**Comment by fjmolinas** [comment]
@kaspar030 I think I addressed your changes let me know if its OK to squash!

File: sys/ztimer/auto_init.c [hunk 8]
```diff
@@ -43,63 +43,23 @@
...
-#define CONFIG_ZTIMER_USEC_WIDTH (32)
-#endif
-
-#ifndef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
-#define CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
-#endif
-
-#ifndef CONFIG_ZTIMER_MSEC_REQUIRED_PM_MODE
-#define CONFIG_ZTIMER_MSEC_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
-#endif
```
Comment (fjmolinas) [inline sys/ztimer/auto_init.c:95]: Ups unintended removed this.

File: sys/include/ztimer.h (hunk 4 above)
Comment (fjmolinas) [reply sys/include/ztimer.h]: done

File: sys/include/ztimer.h (hunk 3 above)
Comment (fjmolinas) [reply sys/include/ztimer.h]: resolved, moved them to `ztimer/include/config.h`

**Comment by fjmolinas** [comment]
@jue89 please take a look as well!

File: sys/include/ztimer/config.h [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+#include "ztimer.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * @brief   Use periph_timer as the base timer for ZTIMER_USEC
+ *
```
Comment (jue89) [inline sys/include/ztimer/config.h:34]: Empty line -> Could be removed

File: sys/include/ztimer.h [hunk 10]
```diff
@@ -513,6 +513,41 @@ extern ztimer_clock_t *const ZTIMER_USEC;
...
 extern ztimer_clock_t *const ZTIMER_MSEC;
 
+/**
+ * @brief   Base ztimer for the microsecond clock (ZTIMER_USEC)
+ *
+ * This ztimer will reference the counter device object at the end of the
+ * chain of ztimer_clock_t for ZTIMER_USEC.
+ *
+ * If the base counter device object's frequency (CONFIG_ZTIMER_USEC_BASE_FREQ)
+ * is not 1Mhz then ZTIMER_USEC will be converted on top of this one. Otherwise
```
Comment (jue89) [inline sys/include/ztimer.h:523]: 1Mhz -> 1MHz

File: sys/include/ztimer/config.h [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ */
+#ifdef MODULE_ZTIMER_PERIPH_RTT
+#  define CONFIG_ZTIMER_MSEC_BASE_FREQ   (RTT_FREQUENCY)
+# else
+#  define CONFIG_ZTIMER_MSEC_BASE_FREQ   (ZTIMER_USEC_BASE_FREQ)
+#endif  /* MODULE_ZTIMER_PERIPH_RTT */
+
+/**
+ * @brief   The minimum pm mode required for the clock to run.
+ *
```
Comment (jue89) [inline sys/include/ztimer/config.h:109]: Empty line -> Could be removed

Furthermore, I would rephrase this to "The minimum pm mode required for ZTIMER_USEC to run."

File: sys/include/ztimer/config.h [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+ */
+#ifndef CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE
+#define CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE ZTIMER_CLOCK_NO_REQUIRED_PM_MODE
+#endif
+/**
+ * @brief   The minimum pm mode required for the clock to run
+ */
```
Comment (jue89) [inline sys/include/ztimer/config.h:116]: Insert an empty line above this block.

Rephrase to "The minimum pm mode required for ZTIMER_MSEC to run."

Comment (fjmolinas) [reply sys/include/ztimer/config.h]: done

File: sys/include/ztimer/config.h (hunk 11 above)
Comment (fjmolinas) [reply sys/include/ztimer/config.h]: done

File: sys/include/ztimer.h (hunk 10 above)
Comment (fjmolinas) [reply sys/include/ztimer.h]: done

File: sys/include/ztimer/config.h (hunk 9 above)
Comment (fjmolinas) [reply sys/include/ztimer/config.h]: done

File: sys/include/ztimer/config.h [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+#define CONFIG_ZTIMER_USEC_BASE_FREQ    (1000000LU)
+#endif
+
+/**
+ * @brief   The frequency of ZTIMER_MSEC_BASE (base ztimer for ZTIMER_MSEC)
+ */
+#ifdef MODULE_ZTIMER_PERIPH_RTT
+#  define CONFIG_ZTIMER_MSEC_BASE_FREQ   (RTT_FREQUENCY)
+# else
+#  define CONFIG_ZTIMER_MSEC_BASE_FREQ   (ZTIMER_USEC_BASE_FREQ)
```
Comment (jue89) [inline sys/include/ztimer/config.h:103]: I did some compile testing.

`ZTIMER_USEC_BASE_FREQ` should be `CONFIG_ZTIMER_USEC_BASE_FREQ`.

Comment (fjmolinas) [reply sys/include/ztimer/config.h]: thanks for the catch!

**Comment by kaspar030** [comment]
please squash!

**Comment by fjmolinas** [comment]
Unrelated test failure re-triggering with no tests [tests/pkg_littlefs2/esp32-wroom-32:gnu](https://ci.riot-os.org/RIOT-OS/RIOT/13990/95e2221003d3e2d2d54bf98bf5d82dc6a4ceb902/output/run_test/tests/pkg_littlefs2/esp32-wroom-32:gnu.txt)

**Comment by fjmolinas** [comment]
All green @kaspar030!

**Comment by fjmolinas** [comment]
Thanks for the review @kaspar030 @jue89!
